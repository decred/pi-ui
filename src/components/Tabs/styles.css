.tab {
  display: flex;
  align-items: center;
  font-size: var(--font-size-normal);
  color: var(--tab-font-color);
  cursor: pointer;
  white-space: nowrap;
}

.tab:hover {
  color: var(--tab-hover-color) !important;
}

.tabHorizontal {
  composes: tab;
  padding-bottom: 0.6rem;
  margin-right: 5rem;
  border-bottom: 0.4rem solid transparent;
}

.tabVertical {
  composes: tab;
  min-height: 3rem;
  border-left: 0.4rem solid transparent;
  padding-left: 0.98rem;
  margin-top: 3rem;
}

.tabDropdownMode {
  composes: tab;
  display: flex;
  border-left: 0;
  padding-left: 0;
  margin-top: 2rem;
}

.tabsNav {
  list-style-type: none;
  display: flex;
  overflow: auto;
}

.wrap {
  flex-wrap: wrap;
}

.tabsNavVertical {
  composes: tabsNav;
  flex-direction: column;
}

.tabCount {
  display: flex;
  align-items: center;
  justify-content: center;
  background: var(--tab-count-background);
  height: 2rem;
  width: 2rem;
  margin-left: 0.5rem;
}

.dropdownArrowClass {
  right: 45px;
  top: 10px;
}

.dropdownListClass {
  margin-top: 3rem;
}

.activeDropdownTabClass {
  margin: 0 1.5rem 0 0;
  padding-left: 1.5rem;
}

.activeDropdownTabWrapper {
  display: flex;
  align-items: center;
  justify-content: flex-start;
  width: 142px;
}
