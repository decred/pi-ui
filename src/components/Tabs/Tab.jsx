import PropTypes from "prop-types";
import React from "react";
import { animated, useSpring } from "react-spring";
import { useTheme, getThemeProperty } from "../../theme";
import { classNames } from "../../utils";
import styles from "./styles.css";

const Tab = ({
  onSelect,
  tabIndex,
  isActive,
  className,
  style,
  label,
  count,
  mode,
  ...props
}) => {
  const { theme } = useTheme();
  const handleOnClick = (e) => {
    e.preventDefault();
    onSelect(tabIndex);
  };
  const activeBorderColor = getThemeProperty(theme, "tab-active-color");
  const defaultBorderColor = getThemeProperty(theme, "tab-default-color");
  const activeBackgroundColor = getThemeProperty(theme, "tab-active-background");
  const defaultBackgroundColor = getThemeProperty(theme, "tab-default-background");
  const activeTextColor = getThemeProperty(theme, "tab-text-active-color");
  const defaultTextColor = getThemeProperty(theme, "tab-text-color");
  const slide = useSpring({
    borderColor: isActive ? activeBorderColor : defaultBorderColor,
    color: isActive ? activeTextColor : defaultTextColor,
    backgroundColor: isActive ? activeBackgroundColor : defaultBackgroundColor,
    duration: 350
  });

  const dropdownMode = mode === "dropdown";
  const vertical = mode === "vertical";

  return (
    <animated.li
      className={classNames(
        vertical
          ? styles.tabVertical
          : dropdownMode
          ? styles.tabDropdownMode
          : styles.tabHorizontal,
        className
      )}
      style={{ ...slide, style }}
      onClick={handleOnClick}
      {...props}>
      <span className={styles.tabLabel} style={style}>
        {label}
      </span>
      {count >= 0 && <span className={styles.tabCount}>{count}</span>}
    </animated.li>
  );
};

Tab.propTypes = {
  onSelect: PropTypes.func,
  tabIndex: PropTypes.number,
  isActive: PropTypes.bool,
  className: PropTypes.string,
  style: PropTypes.object,
  label: PropTypes.node,
  count: PropTypes.node,
  mode: PropTypes.oneOf(["horizontal", "vertical", "dropdown"])
};

Tab.defaultProps = {
  mode: "horizontal"
};

export default Tab;
