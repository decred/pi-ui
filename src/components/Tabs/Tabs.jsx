import PropTypes from "prop-types";
import React, { useCallback, useMemo } from "react";
import { animated, useTransition } from "react-spring";
import { classNames } from "../../utils";
import styles from "./styles.css";
import Dropdown from "../Dropdown/Dropdown.jsx";

const TabDropdownTrigger = ({
  onClick,
  open,
  ArrowComponent,
  childrenTabs,
  activeTabIndex
}) => (
  <div className={styles.activeDropdownTabWrapper}>
    {React.Children.map(childrenTabs, (child, index) => {
      if (index === activeTabIndex) {
        return React.cloneElement(child, {
          onClick: onClick,
          className: styles.activeDropdownTabClass,
          isActive: true,
          mode: "dropdown"
        });
      }
    })}
    <ArrowComponent onClick={onClick} open={open} />
  </div>
);

const Tabs = ({
  onSelectTab,
  activeTabIndex,
  style,
  className,
  children,
  wrap,
  mode,
  ...props
}) => {
  const dropdownMode = mode === "dropdown";
  const vertical = mode === "vertical" || dropdownMode;
  const renderChildrenTabs = useCallback(() => {
    return React.Children.toArray(children)
      .filter(Boolean)
      .map((child, index) => {
        return React.cloneElement(child, {
          "data-testid": `tab-${index}`,
          onSelect: onSelectTab,
          tabIndex: index,
          isActive: index === activeTabIndex,
          mode
        });
      });
  }, [children, activeTabIndex, mode]);

  const tabs = useMemo(
    () => (
      <ul
        className={classNames(
          vertical ? styles.tabsNavVertical : styles.tabsNav,
          wrap && styles.wrap,
          className
        )}
        style={style}
        {...props}>
        {renderChildrenTabs()}
      </ul>
    ),
    [vertical, wrap, className, props, renderChildrenTabs]
  );

  const getActiveChild = (props) => {
    return (
      <TabDropdownTrigger
        {...props}
        childrenTabs={children}
        activeTabIndex={activeTabIndex}
      />
    );
  };

  const transitions = useTransition(activeTabIndex, null, {
    from: { opacity: 0 },
    enter: { opacity: 1 },
    leave: { opacity: 0 },
    config: { duration: 350 }
  });

  return (
    <>
      {dropdownMode ? (
        <Dropdown
          customDropdownTrigger={getActiveChild}
          closeOnOutsideClick={true}
          className={className}
          itemsListClassName={classNames(styles.dropdownListClass)}>
          {tabs}
        </Dropdown>
      ) : (
        tabs
      )}
      {transitions.map(({ item, key, props }) => {
        return (
          item === activeTabIndex && (
            <animated.div key={key} style={props}>
              {children[item] && children[item].props.children}
            </animated.div>
          )
        );
      })}
    </>
  );
};

TabDropdownTrigger.propTypes = {
  onClick: PropTypes.func,
  open: PropTypes.bool,
  ArrowComponent: PropTypes.func,
  childrenTabs: PropTypes.node,
  activeTabIndex: PropTypes.number
};

Tabs.propTypes = {
  className: PropTypes.string,
  style: PropTypes.object,
  onSelectTab: PropTypes.func.isRequired,
  activeTabIndex: PropTypes.number.isRequired,
  children: PropTypes.node.isRequired,
  wrap: PropTypes.bool,
  mode: PropTypes.oneOf(["horizontal", "vertical", "dropdown"])
};

Tabs.defaultProps = {
  wrap: false,
  mode: "horizontal"
};

export default Tabs;
