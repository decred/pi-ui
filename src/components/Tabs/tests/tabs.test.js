import React from "react";
import { create } from "react-test-renderer";
import {
  defaultLightTheme,
  ThemeProvider,
  DEFAULT_LIGHT_THEME_NAME,
} from "../../../theme";
import { render, fireEvent, waitFor } from "@testing-library/react";
import Tabs from "../Tabs";
import Tab from "../Tab";

describe("Tabs Component", () => {
  test("Matches the snapshot", () => {
    const tabs = create(
      <ThemeProvider
        themes={{ [DEFAULT_LIGHT_THEME_NAME]: defaultLightTheme }}
        defaultThemeName={DEFAULT_LIGHT_THEME_NAME}>
        <Tabs onSelectTab={jest.fn()} activeTabIndex={0}>
          <Tab label="tab1" count={1}>
            <div>test1</div>
          </Tab>
          <Tab label="tab2" count={4}>
            <div>test2</div>
          </Tab>
        </Tabs>
      </ThemeProvider>
    );
    expect(tabs.toJSON()).toMatchSnapshot();
  });

  test("Changing tabs works properly", async () => {
    // create TabsContainer helper component to be able
    // to serve activeTabIndex state parameter to Tabs
    const TabsContainer = () => {
      const [activeTabIndex, setActiveTabIndex] = React.useState(0);

      return (
        <ThemeProvider
          themes={{ [DEFAULT_LIGHT_THEME_NAME]: defaultLightTheme }}
          defaultThemeName={DEFAULT_LIGHT_THEME_NAME}>
          <Tabs onSelectTab={setActiveTabIndex} activeTabIndex={activeTabIndex}>
            <Tab label="tab1" count={1}>
              <div>test1</div>
            </Tab>
            <Tab label="tab2" count={4}>
              <div>test2</div>
            </Tab>
          </Tabs>
        </ThemeProvider>
      );
    };
    const { getByTestId, queryByText, getByText, queryAllByText } = render(
      <TabsContainer />
    );
    expect(queryByText(/test1/i)).toBeTruthy();
    expect(queryByText(/test2/i)).toBeFalsy();
    fireEvent.click(getByTestId("tab-1"));

    // wait until `test2` content appears entirely
    await waitFor(() =>
      expect(getByText("test2").parentNode.style.opacity).toBe("1")
    );
    expect(queryByText(/test1/i)).toBeFalsy();
    expect(queryAllByText(/test2/i)).toBeTruthy();
  });
});
