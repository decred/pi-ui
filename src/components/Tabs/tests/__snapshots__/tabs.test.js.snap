// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Component Matches the snapshot 1`] = `
Array [
  <ul
    className="tabsNav"
  >
    <li
      className="tabHorizontal"
      data-testid="tab-0"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "borderColor": "rgba(46, 216, 163, 1)",
          "color": "rgba(9, 20, 64, 1)",
          "duration": 350,
          "style": undefined,
        }
      }
    >
      <span
        className="tabLabel"
      >
        tab1
      </span>
      <span
        className="tabCount"
      >
        1
      </span>
    </li>
    <li
      className="tabHorizontal"
      data-testid="tab-1"
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "transparent",
          "borderColor": "rgba(255, 255, 255, 1)",
          "color": "rgba(61, 88, 115, 1)",
          "duration": 350,
          "style": undefined,
        }
      }
    >
      <span
        className="tabLabel"
      >
        tab2
      </span>
      <span
        className="tabCount"
      >
        4
      </span>
    </li>
  </ul>,
  <div
    style={
      Object {
        "opacity": 1,
        "position": "absolute",
      }
    }
  >
    <div>
      test1
    </div>
  </div>,
]
`;
